vassal-app/src/main/java/VASSAL/build/module/KeyNamer.java

```diff
@@ -0,0 +1,111 @@
+ * License (LGPL) as published by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public
+ * License along with this library; if not, copies are available
+ * at http://www.opensource.org.
+ */
+package VASSAL.build.module;
+
+import java.awt.event.KeyEvent;
+
+import javax.swing.KeyStroke;
+
+import org.apache.commons.lang3.StringUtils;
+
+import VASSAL.build.Buildable;
+import VASSAL.i18n.Resources;
+import VASSAL.tools.swing.SwingUtils;
+
+
+/**
+ * Translates key codes into human-readable strings (e.g. "Backspace"). 
+ * 
+ * Exists as a buildable (buildFile element) so that it can be overridden by custom classes
+ */
+public class KeyNamer implements Buildable {
+  @Override
+  public void build(org.w3c.dom.Element e) {
+  }
+
+  @Override
+  public org.w3c.dom.Element getBuildElement(org.w3c.dom.Document doc) {
+    return doc.createElement(getClass().getName());
+  }
+
+  @Override
+  public void addTo(Buildable b) {
+  }
+
+  @Override
+  public void add(Buildable b) {
+  }  
+
+  /**
+   * Implements VASSAL's plain text representation of a KeyStroke.
+   * @param k - a keystroke (e.g. VK_BACKSPACE) to be rendered as a string
+   * @return  - a human-readable name for the keystroke (e.g. "Backspace")
+   */
+  public static String getKeyString(KeyStroke k) {
+    if (k == null) {
+      return null;
+    }
+
+    StringBuilder sb = new StringBuilder();
+    int code = k.getKeyCode();
+    switch (code) {
+    // The addition of underscores screws up these names
+    case KeyEvent.VK_ADD:
+      sb.append(Resources.getString("Keys.numplus")); //$NON-NLS-1$ //$NON-NLS-2$
+      break;
+    case KeyEvent.VK_SUBTRACT:
+      sb.append(Resources.getString("Keys.numminus")); //$NON-NLS-1$ //$NON-NLS-2$
+      break;
+    // More compact name, also commonly printed keys  
+    case KeyEvent.VK_PAGE_UP:
+      sb.append(Resources.getString("Keys.pgup")); //$NON-NLS-1$ //$NON-NLS-2$
+      break;
+    case KeyEvent.VK_PAGE_DOWN:
+      sb.append(Resources.getString("Keys.pgdn")); //$NON-NLS-1$ //$NON-NLS-2$
+      break;
+    // Non-Americans were (rightly) commenting about this
+    case KeyEvent.VK_OPEN_BRACKET:
+      sb.append(Resources.getString("Keys.bropen")); //$NON-NLS-1$ //$NON-NLS-2$
+      break;
+    case KeyEvent.VK_CLOSE_BRACKET:
+      sb.append(Resources.getString("Keys.brclose")); //$NON-NLS-1$ //$NON-NLS-2$
+      break;
+    default:
+      break;
+    }
+    if (sb.toString().isEmpty() || sb.toString().contains("Keys.")) { //$NON-NLS-1$
+      sb.append(KeyEvent.getKeyText(code)); 
+    }
+
+    final int mods = SwingUtils.genericToSystem(k).getModifiers();
+    if ((mods & KeyEvent.SHIFT_DOWN_MASK) > 0) {
```

Comment:
Technically you could have a mask on the high bit and get a result less than 0 indicating that the masked bit was on, because Java foolishly lacks unsigned integral types... So these should be `!= 0`.